src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -108,56 +137,84 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
             properties = selectedInstanceProps.toIndexedSeq.filter(p =>
               !p.startsWith("node.") && !p.startsWith("edge.") && !p.startsWith("metadata.") &&
                 p != "startNode" && p != "endNode" && p != "space" && p != "externalId" && p != "type")
-        ))
+        )
+      )
       .toSeq
 
-    val instanceSyncRequest = InstanceSyncRequest(
+    InstanceSyncRequest(
       `with` = Map("sync" -> tableExpression),
-      cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None,
+      cursors = currentCursor.flatMap(c => Some(Map("sync" -> c))),
       select = Map("sync" -> SelectExpression(sources = sourceReference)),
       includeTyping = Some(true)
     )
+  }
 
-    Seq(syncOut(instanceSyncRequest, selectedColumns))
+  private def syncOut(
+      hasDataFilter: Boolean,
+      filters: Array[Filter],
+      selectedColumns: Array[String],
+      currentCursor: Option[String])(
+      implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] = {
+
+    val synced: IO[(ItemsWithCursor[InstanceDefinition], Boolean)] =
+      syncWithCursor(hasDataFilter, filters, selectedColumns, currentCursor)
+    val stream: Stream[IO, ProjectedFlexibleDataModelInstance] = Stream.eval(synced).flatMap {
+      case (ItemsWithCursor(items, nextCursor), continueWithHasData) =>
+        val projectedItems =
+          items.filter(clientsideHasDataFilter).map(toProjectedInstance(_, nextCursor, selectedColumns))
+
+        val nextStream = nextCursor match {
+          case Some(c) if items.nonEmpty && items.last.lastUpdatedTime < terminationTimeStamp =>
+            syncOut(continueWithHasData, filters, selectedColumns, Some(c))
+          case _ => Stream.empty
+        }
+        Stream.emits(projectedItems) ++ nextStream
+    }
+    stream
   }
-  // scalastyle:on cyclomatic.complexity
-
-  private def syncOut(syncRequest: InstanceSyncRequest, selectedProps: Array[String])(
-      implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] =
-    Stream.eval {
-      syncWithCursor(syncRequest).map { sr =>
-        val items = sr.items
-        val nextCursor = sr.nextCursor
-        val next =
-          (nextCursor, items.nonEmpty && items.last.lastUpdatedTime < terminationTimeStamp) match {
-            case (Some(cursor), true) =>
-              syncOut(syncRequest.copy(cursors = Some(Map("sync" -> cursor))), selectedProps)
-            case _ => fs2.Stream.empty
-          }
 
-        val projected = items.map(toProjectedInstance(_, nextCursor, selectedProps))
-        fs2.Stream.emits(projected) ++ next
-      }
-    }.flatten
+  private def clientsideHasDataFilter(item: InstanceDefinition): Boolean = {
+    val values = item.properties.getOrElse(Map.empty).values.flatMap(_.values).fold(Map.empty)(_ ++ _)
+    values.nonEmpty &&
+    mandatoryViewFields.forall { field =>
```

Comment:
A little unsure about this part, https://api-docs.cognite.com/20230101/tag/Instances/operation/syncContent#!path=with/property%20name*/0/nodes/filter/1/10/hasData&t=request and docs around don't say a lot about
- in /sync endpoint are all properties returned or only changed (seems like all)
- should it be an error if some of mandatory fields are present but some are missing?
- what happens if hasData filter isn't set in request? Would that mean getting unrelated views' changes?
- could we instead filter by view space+externalId?

Maybe also add a sanity assert or test that if server-side hasData filtering is applied then clientside filter agrees?
Also if not too hard to set up a unit test that compares clientside <> server-side filtering logic would be nice